designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+
+        # Do the build
+        output_folder = make_build_folder(resource)
+        builder.build(resource.Code, resource.runtime, output_folder)
+
+
+We will keep the implementation of build agnostic of the resource type. This opens up the future possibility of adding
+build actions to any resource types, not just Lambda functions. Initially we will start by supporting only the
+resource types ``AWS::Serverless::Function`` and ``AWS::Lambda::Function``.
+
+Build Folder
+~~~~~~~~~~~~
+Default Location: ``$PKG_ROOT/build/``
+
+By default, we will create a folder called ``build`` right next to where the SAM template is located.
+This will contain the built artifacts for each resource. Customers can always override this folder location.
+
+Built artifacts for each resource will be stored within a folder named with the LogicalID of the resource.
+This allows us to build separate zip files for each Lambda, so users can update one Lambda without triggering an update
+on another. The same model will work for building other non-Lambda resources.
+
+*Advantages:*
+
+* Extensible to other resource types
+* Supports parallel builds for each resource
+* Aligned with a CloudFormation stack
+
+*Disadvantages:*
+
+* Too many build folders, and hence zip files, to manage.
+* Difficult to share code between all Lambdas.
+
+::
+
+    $PKG_ROOT/build/
+        artifacts.json (not for MVP)
+        MyFunction1/
+             .... <build artifacts>
+
+          MyFunction2/
+             ... <build artifacts>
+
+          MyApiGw/
+            ... <build artifacts>
+
+          MyECRContainer/
+             ... <build artifacts>
+
+
+Future Extensions
+^^^^^^^^^^^^^^^^^
+In the future, we will change the limitation around each folder being named after the resource's LogicalID.
+Instead, we will support an artifacts.json file that will map Lambda function resource's LogicalId to the path to a
+folder that contains built artifacts for this function. This allows us to support custom build systems that use
+different folder layout.
+
+A well-known folder structure also helps “sam local” and “sam package” commands to automatically discover the
+built artifacts for each Lambda function and package it.
+
+Stable Builds
+~~~~~~~~~~~~~
+A build is defined to be stable if the built artifacts changes if and only if the contents or metadata
+(ex: timestamp, ownership) on source files changes. This is an important attribute of a build system. Since SAM CLI
+relies on 3rd party package managers like NPM to do the heavy lifting, we can only provide a "best effort" service here.
+By running ``sam build`` on a build system that creates a new environment from scratch
+(ex: Travis/CircleCI/CodeBuild/Jenkins etc), you can achieve truly stable builds. For more information on why this
+is important, refer to Debian's guide on `reproducible builds <https://reproducible-builds.org/>`_.
+
+
+SAM CLI does the following to produce stable builds:
+
+#. Clean build folder on every run
```

Comment:
should we clean it, only if there is change in source?

Reply:
It will give better performance. But it also makes the build implementation complicated. We can start here and then improve this in future